File: src/PwrDrvr.MicroApps.Deployer/Lambda/GatewayInfo.cs
using System.Threading.Tasks;
using Amazon.ApiGatewayV2;
using Amazon.ApiGatewayV2.Model;

namespace PwrDrvr.MicroApps.Deployer.Lambda {
  public class GatewayInfo {
    public static async Task<string> GetAPIID(AmazonApiGatewayV2Client apigwy) {
      return (await GetAPI(apigwy)).ApiId;
    }

    public static async Task<Api> GetAPI(AmazonApiGatewayV2Client apigwy) {
      var apis = await apigwy.GetApisAsync(new GetApisRequest() {
        MaxResults = "100",
      });
      // TODO: Handle pagination
      foreach (var api in apis.Items) {
        if (api.Name == "microapps-apis") {
          return api;
        }
      }

      return null;
    }
  }
}

File: src/microapps-deployer/src/lib/GatewayInfo.ts
import * as apigwy from '@aws-sdk/client-apigatewayv2';

export default class GatewayInfo {
  // public static async GetAPIID(apigwyClient: apigwy.ApiGatewayV2Client): Promise<string> {
  //   return (await GatewayInfo.GetAPI(apigwyClient))?.ApiId as string;
  // }

  public static async GetAPI(
    apigwyClient: apigwy.ApiGatewayV2Client,
  ): Promise<apigwy.Api | undefined> {
    let apis: apigwy.GetApisCommandOutput | undefined;
    do {
      const optionals =
        apis?.NextToken !== undefined
          ? { NextToken: apis.NextToken }
          : ({} as apigwy.GetApisCommandInput);
      apis = await apigwyClient.send(
        new apigwy.GetApisCommand({
          MaxResults: '100',
          ...optionals,
        }),
      );

      if (apis === undefined) {
        throw new Error('GetApisCommand unexpectedly returned undefined');
      }
      if (apis.Items === undefined) {
        continue;
      }

      // Loop through and find our item, it it is here
      for (const api of apis.Items) {
        // FIXME: Load the api name from an env var
        if (api.Name === 'microapps-apis') {
          return api;
        }
      }
    } while (apis.NextToken !== undefined);

    return undefined;
  }
}
